test/common/config/metadata_test.cc

```diff
@@ -60,7 +60,7 @@ class TypedMetadataTest : public testing::Test {
 
   struct Bar : public TypedMetadata::Object {};
 
-  class FooFactory : public TypedMetadataFactory::TypedMetadataFactory {
+  class FooFactory : public Envoy::Config::TypedMetadataFactory {
```

Comment:
Why is this needed? It seems lexically obvious we are in a namespace where TypedMetadataFactory should be usable without qualification. If this cannot work I wonder what other things will fail?

Reply:
From this commit log;
> TypedMetadataFactory::TypedMetadataFactory remains too abstract, and didn't have base class methods name or parse.

We are quite confused as well, this is the error output of the original source;
```
ERROR: C:/workspace/envoy/test/common/config/BUILD:261:1: C++ compilation of rule '//test/common/config:metadata_test_lib_internal_only' failed (Exit 2)
test/common/config/metadata_test.cc(63): error C2516: 'Envoy::Config::TypedMetadataFactory::{ctor}': is not a legal base class
test/common/config/metadata_test.cc(65): error C3668: 'Envoy::Config::`anonymous-namespace'::TypedMetadataTest::FooFactory::name': method with override specifier 'override' did not override any base class methods
test/common/config/metadata_test.cc(68): error C3668: 'Envoy::Config::`anonymous-namespace'::TypedMetadataTest::FooFactory::parse': method with override specifier 'override' did not override any base class methods
test/common/config/metadata_test.cc(54): error C2664: 'Envoy::Registry::InjectFactory<Envoy::Config::TypedMetadataFactory>::InjectFactory(const Envoy::Registry::InjectFactory<Envoy::Config::TypedMetadataFactory> &)': cannot convert argument 1 from 'Envoy::Config::`anonymous-namespace'::TypedMetadataTest::FooFactory' to 'Base &'
        with
        [
            Base=Envoy::Config::TypedMetadataFactory
        ]
.\test/test_common/registry.h(17): note: see declaration of 'Envoy::Registry::InjectFactory<Envoy::Config::TypedMetadataFactory>::InjectFactory'
```

Things that are confusing, why promote TypedMetadataFactory:: to Envoy::Config:: in these contexts? Why the anon namespace? Suggestions and observations welcome.

Reply:
Does it work if you just extend `TypedMetadataFactory` instead of `TypedMetadataFactory::TypedMetadataFactory`? The latter seems to be nonsense, since there's no TypedMetadataFactory namespace in the code.

Reply:
@zuercher agreed, that was the issue. Will push the fix first thing Friday.